ls1intum/Artemis
# `Exam mode`: Improve client-side exam import validation

### Checklist
#### General



#### Client

### Motivation and Context


When importing an exam or an exercise group, programming exercises are checked for duplicated titles and short names on the server side (introduced in https://github.com/ls1intum/Artemis/pull/6877). In this PR a client side check is introduced. 

### Description

When first selecting an exam / exercise group to import, duplicated titles of programming exercises are removed and replaced by a placeholder.  It is then possible to change the title and short name of any given exercise. In case two or more programming exercises have the same title or short name, the corresponding input fields are marked red. Note that only selected programming exercises are taken account. This means when an exercise is unselected it is no longer considered to be duplicated. When not all programming exercise titles and short names are unique, the exam will not be saved. 

### Steps for Testing

Prerequisites:
- 1 Instructor
- Exam with programming exercises

1. Navigate to Course Management to a course
3. Go to Exams > Import an exam  (or for an existing exam: > Import Exercise Group)
4. Import an exam with duplicated titles (if on ts1: you can import `Duplicated Titles` Exam from `Random Course`, otherwise you can create with different programming exercises and after the exercise creation edit the titles so they match)
5. Click on Import and make sure that the duplicated titles are removed
6. Try changing the titles and short names of exercises and make sure duplicates are marked correctly
7. Also try unselecting an exercises => the exercise should no longer be considered a duplicate
8. Fill out visible from, start and end of working time
9. Click on import when there are duplicates  and see the alert message appear, the exam should not be saved

### Review Progress






#### Performance Review

#### Code Review
#### Manual Tests

### Test Coverage







| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [exam-exercise-import.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5250-5/artifact/shared/Coverage-Report-Client-Tests/app/exam/manage/exams/exam-exercise-import/index.html) | 100% | ✅ |

### Screenshots


Duplicated title replaced by a placeholder after selecting exam to import:
![image](https://github.com/ls1intum/Artemis/assets/73339358/05b14a91-03ba-4958-a6f1-92b3d27605e3)

Duplicated titles and short names are marked red after editing input:
![image](https://github.com/ls1intum/Artemis/assets/73339358/6b43daf9-831b-41d6-a1fa-fd4c54c9ba93)

Importing an exam with duplicated titles, checking for dulicated titles & short names:

https://github.com/ls1intum/Artemis/assets/73339358/ad02c4d7-6c7e-41ed-a032-6a48dfb4e8d5
